### ISSUE TITLE
binding vs. uid API methods

### ISSUE THREAD

**Issue by DmitrySoshnikov** [babel/babel#2101]
The destructuring generates for the:

```js
var {foo} = require('bar');
```

the following code:

```js
var _require = require('bar');
var foo = _require.foo;
```

The `_require` var goes to the `uids` instead of `bindings`, however, `scope.hasBinding` returns `true` since considers uids.

Then our "after" plugin removes the `var _require = require('bar');` node, in order to inline the call. However, `scope.removeBindings` doesn't affect it, since it's not in bindings, but in uids.

We should probably add such generated vars to bindings as well, or to expose `scope.removeUid` method. And also `scope.hasBinding` (returns `true`) is kind of inconsistent after  `scope.removeBinding` (which does nothing in this case).

**Comment by voideanvalue** [comment]
I think `_require` getting added to `scope.uids` but not `scope.bindings` is fine (unless we were to add a `binding` such that `binding.kind` is `"temporary"` or something else that reflects the fact that it was injected as part of a transform). We could either add a `scope.removeUid` and increase the API surface or just make `scope.removeBinding` remove the `uid` if the parameter passed in isn't a `binding` (justifiable since `scope.hasBinding` is also overloaded in a similar way and `uids` & `bindings` won't clash based on how `uids` are generated).

If we added `_require` to both `scope.bindings` and `scope.uids`, `scope.removeBinding` would then have to clean up for consistency.

**Comment by DmitrySoshnikov** [comment]
Yeah, probably everything that generates `var` statement should be considered as a binding (since it's actually a binding). Or, if we really want to distinguish them, then `scope.removeBinding` should remove it (but in this case a uid becomes even more indistinguishable from a binding, and is just an implementation detail).

**Comment by sebmck** [comment]
It doesn't get attached to `bindings` since we don't have enough information about the potential binding to track it. ie. a binding needs to have a reference in the AST and on creation of UIDs it might not even be present.

If scope tracking has already been performed on a branch then on each traversal it's skipped which means that declarations inserted after the initial scope tracking aren't updated either. I've experimented with ways to somehow keep it in sync but it quickly delved into madness where it was causing way too many invalidations and increased traversal significantly. There's likely a way to optimise this and hook into nodes being removed/added to reflow the scope tracking.

Overloading `scope.removeBinding` seems like the right solution, the current UID being kinda recognised as a binding is hacky but necessary because of these limitations.

**Comment by DmitrySoshnikov** [comment]
OK, sounds good, thanks!

**Comment by DmitrySoshnikov** [comment]
Also, could you publish this version please?

**Comment by sebmck** [comment]
Done, published 5.8.14.

**Comment by DmitrySoshnikov** [comment]
Thanks!
